src/election.ts

```diff
@@ -0,0 +1,165 @@
+              public readonly name: string,
+              public readonly ttl: number = 60) {
+    this.namespace = namespace.namespace(this.getPrefix());
+    this.lease = this.namespace.lease(ttl);
+  }
+
+  public async ready() {
+    const leaseId = await this.lease.grant();
+
+    if (!this.isReady) {
+      this._leaseId = leaseId;
+    }
+  }
+
+  public async campaign(value: any) {
+    this.throwIfNotReady();
+    const result = await this.namespace
+      .if(this.leaseId, 'Create', '==', 0)
+      .then(this.namespace.put(this.leaseId).value(value).lease(this.leaseId))
+      .else(this.namespace.get(this.leaseId))
+      .commit();
+
+    this._leaderKey = `${this.getPrefix()}${this.leaseId}`;
+    this._leaderRevision = result.header.revision;
+    this._isLeader = true;
+
+    if (!result.succeeded) {
+      try {
+        const kv = result.responses[0].response_range.kvs[0];
+        this._leaderRevision = kv.create_revision;
+        await this.proclaim(value);
+      } catch (error) {
+        await this.resign();
+        throw error;
+      }
+    }
+
+    try {
+      await this.waitForElected();
+    } catch (error) {
+      this._isLeader = false;
+      throw error;
+    }
+  }
+
+  public async proclaim(value: any) {
+    this.throwIfNotReady();
+
+    if (!this._isLeader) {
+      throw Election.notLeaderError;
+    }
+
+    const r = await this.namespace
+      .if(this.leaseId, 'Create', '==', this._leaderRevision)
+      .then(this.namespace.put(this.leaseId).value(value).lease(this.leaseId))
+      .commit();
+
+    if (!r.succeeded) {
+      this._leaderKey = '';
+      throw Election.notLeaderError;
+    }
+  }
+
+  public async resign() {
+    this.throwIfNotReady();
+
+    if (!this.isLeader) {
+      return;
+    }
+
+    await this.namespace
+      .if(this.leaseId, 'Create', '==', this._leaderRevision)
+      .then(this.namespace.delete().key(this.leaseId))
+      .commit();
+
+    this._leaderKey = '';
+    this._leaderRevision = '';
+    this._isLeader = false;
+  }
+
+  public async getLeader() {
+    const result = await this.namespace.getAll().sort('Create', 'Ascend').keys();
+    if (result.length === 0) {
+      throw Election.noLeaderError;
+    }
+    return `${this.getPrefix()}${result[0]}`;
+  }
+
+  public getPrefix() {
+    return `${Election.prefix}/${this.name}/`;
+  }
+
+  private async waitForElected() {
+    // find last create before this
+    const lastRevision = new BigNumber(this.leaderRevision).minus(1).toString();
+    const result = await this.namespace.getAll().maxCreateRevision(lastRevision).keys();
+
+    // no one before this, elected
+    if (result.length === 0) {
+      return;
+    }
+
+    const lastKey = result[0];
+    const watcher = await this.namespace.watch().key(lastKey).create();
+    const deleteOrError = new Promise(async (resolve, reject) => {
+      // waiting for deleting of that key
+      watcher.on('delete', resolve);
+      watcher.on('error', reject);
```

Comment:
as a point of discussion (I recently had to implement this exact functionality as an extension to this lib for work last month) -- I elected to also listen to the `disconnected` event on the watcher so that the election can react to being disconnected for too long as at that point the application may want to take action given that the election cannot know for sure where it is in the line of succession until the network partition ends. The Golang implementation does not make any distinction between particular types of `WatchResponse` errors but treats them all the same. In any case I'm curious what people's thoughts are on this one?

Reply:
`disconnected` should be listen, I miss this exception. But I prefer throw those errors to application layer and let users to decision how to react with. Election is a fundamental instruction do no more thing but election IMO.

Reply:
agree that application should get to decide what to do. In my work implementation I propagate all failures back up the application -- including losing the lease, not just watcher failures. That said, I chose to call `resign` in my implementation if there is an unrecoverable watcher failure (or unrecoverable lease failure) because `resign` clears the internal state of the election client such that when/if the application decides to `campaign` again, it is a clean "do-over" from the failure state. If we don't call `resign`, then while calling `campaign` again from the application should basically just attempt to re-acquire the lease and then update election state if it succeeds, it felt cleaner to me that the since a lost watcher/lost lease essentially compromises the client's view of the line-of-succession, that's logically equivalent to a resignation as in either case the client can no longer know where it stands until it reconnects.

Therefore, if we assume that the lease is unrecoverable and the watchers are unrecoverable, when the partition ends we are just at the back of the line again either way. This may just be a personal preference for managing the internal state machine, but it felt cleaner and it didn't seem to introduce anything the application couldn't control.

To be clear, we would only `resign` internally (as the Golang impl does) on an _unrecoverable_ network failure (lost lease or permanent watcher error) not a transient failure like a `keepAliveFailure` or watcher `disconnect`.

Food for thought I guess...

Reply:
You don't necessarily need to listen to `disconnected`, aside from handling timeouts, since when we reconnect we'll ask etcd to replay any changes that have happened since the time when we disconnected. So from an application pov it'll just be in the "waiting" state for longer than normal but otherwise should be entirely consistent. Being able to time out would be nice, maybe allow an optional `timeout` parameter to be passed in `campaign()`.

I agree that we should attempt to `resign` in all failure cases.